packages/core/js-client/src/index.ts

```diff
@@ -13,157 +13,236 @@
-
-const fetchWorkerCode = () => fetchResource('@fluencelabs/marine-worker', '/dist/browser/marine-worker.umd.cjs').then(res => res.text());
-const fetchMarineJsWasm = () => fetchResource('@fluencelabs/marine-js', '/dist/marine-js.wasm').then(res => res.arrayBuffer());
-const fetchAvmWasm = () => fetchResource('@fluencelabs/avm', '/dist/avm.wasm').then(res => res.arrayBuffer());
-
-const createClient = async (relay: RelayOptions, config: ClientConfig): Promise<IFluenceClient> => {
-    const marineJsWasm = await fetchMarineJsWasm();
-    const avmWasm = await fetchAvmWasm();
-    
-    const marine = new MarineBackgroundRunner({
-        async getValue() {
-            if (isNode) {
-                const require = module.createRequire(import.meta.url);
-                const pathToThisFile = path.dirname(url.fileURLToPath(import.meta.url));
-                const pathToWorker = require.resolve('@fluencelabs/marine-worker');
-                const relativePathToWorker = path.relative(pathToThisFile, pathToWorker);
-                return new Worker(relativePathToWorker);
-            } else {
-                const workerCode = await fetchWorkerCode();
-                return BlobWorker.fromText(workerCode)
-            }
-        },
-        start() {
-            return Promise.resolve(undefined);
-        },
-        stop() {
-            return Promise.resolve(undefined);
-        },
-    }, {
-        getValue() {
-            return marineJsWasm;
-        }, start(): Promise<void> {
-            return Promise.resolve(undefined);
-        }, stop(): Promise<void> {
-            return Promise.resolve(undefined);
-        }
-    }, {
-        getValue() {
-            return avmWasm;
-        }, start(): Promise<void> {
-            return Promise.resolve(undefined);
-        }, stop(): Promise<void> {
-            return Promise.resolve(undefined);
+
+import module from "module";
+import path from "path";
+import process from "process";
+import url from "url";
+
+import type {
+  ClientConfig,
+  ConnectionState,
+  IFluenceClient,
+  RelayOptions,
+} from "@fluencelabs/interfaces";
+
+// "threads" package has broken type definitions in package.json. This is the workaround.
+import { BlobWorker, Worker } from "../node_modules/threads/dist/index.js";
+
+import { ClientPeer, makeClientPeerConfig } from "./clientPeer/ClientPeer.js";
+import { callAquaFunction } from "./compilerSupport/callFunction.js";
+import { registerService } from "./compilerSupport/registerService.js";
+import { fetchResource } from "./fetchers/index.js";
+import { MarineBackgroundRunner } from "./marine/worker/index.js";
+import { doRegisterNodeUtils } from "./services/NodeUtils.js";
+
+const isNode =
+  typeof process !== "undefined" && process.release.name === "node";
+
+const fetchWorkerCode = () => {
+  return fetchResource(
+    "@fluencelabs/marine-worker",
+    "/dist/browser/marine-worker.umd.cjs",
+  ).then((res) => {
+    return res.text();
+  });
```

Comment:
I don't know if it's at all reasonable to forbid using `.then` but I would say that probably in all cases async function can be used instead. Here it would be totally appropriate to use it

Reply:
the same goes for the functions below

Reply:
fixed